File: basic-go/leetcode/list/reverse_list.go
package list

// 反转链表

// 朴素算法： 使用双指针 p 和 q 分别指向需要反转的节点以及其前一个节点，遍历整个链表直到链表尾部
// 时间复杂度: O(N)，需要遍历每个节点
// 空间复杂度： O(1)
func reverseList(head *ListNode) *ListNode {
	var prev *ListNode
	for head != nil {
		prev, head, head.Next = head, head.Next, prev
	}
	return prev
}

// 穿针引线法
func reverseInsertion(head *ListNode) *ListNode {
	dummy := &ListNode{0, head}
	prev, curr := dummy, head
	for curr.Next != nil {
		next := curr.Next
		prev.Next, curr.Next, next.Next = next, next.Next, curr
	}
	return dummy.Next
}

// 递归实现： 考虑将链表分为当前节点和已经反转完成的链表，则只需要将当前链表和已经反转完成的链表链接起来即可
// 时间复杂度： O(N)
// 空间复杂度： O(N)
func recursiveReverseList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	newHead := recursiveReverseList(head.Next)
	head.Next.Next = head
	head.Next = nil
	return newHead
}

// 反转从位置 m 到位置 n 的链表

// 穿针引线法：将每一个需要反转的节点反转到需要反转的部分的起始位置
//    		2->5->4->3
// step1	5->2->4->3
// step2	4->5->2->3
// step3	3->4->5->2
func reverseBetween(head *ListNode, left int, right int) *ListNode {
	dummy := &ListNode{0, head}
	prev := dummy
	for i := 0; i < left; i++ {
		prev = prev.Next
	}
	curr := prev.Next
	// curr 指向需要反转部分的第一个节点，随着后续节点的插入会移动到需要反转的部分的尾部
	for i := left; i <= right; i++ {
		next := curr.Next
		prev.Next, curr.Next, next.Next = next, next.Next, curr
	}
	return dummy.Next
}

File: basic-algorithm/example/src/leetcode/list/ReverseList.java
package leetcode.list;

public class ReverseList {

    /**
     * 两两交换链表中相邻的节点，返回交换后的链表
     */
    public ListNode swapPairs(ListNode head){

        ListNode dummy = new ListNode(-1, head);
        ListNode prev = dummy;
        while (prev.next != null && prev.next.next != null){
            ListNode end = prev.next.next;
            prev.next.next = end.next;
            end.next = prev.next;
            prev.next = end;

            prev = prev.next.next;
        }
        return dummy.next;
    }

    /**
     *  给定单链表头节点 head，返回反转后的链表
     *
     *  穿针引线法：依次将每个节点移动到头节点前面
     */
    public ListNode reverseList(ListNode head){

        if (head == null){
            return null;
        }

        ListNode dummy = new ListNode(-1, head);

        while (head.next != null){
            ListNode reverse = head.next;
            head.next = reverse.next;
            reverse.next = dummy.next;
            dummy.next = reverse;
        }
        return dummy.next;
    }

    /**
     * 反转指定位置之间的链表
     *
     * 穿针引线法
     */
    public ListNode reverseBetween(ListNode head, int left, int right){
        ListNode dummy = new ListNode(-1, head);
        ListNode pre = dummy;
        int i = 0;
        while (i < left - 1){
            i++;
            pre = pre.next;
        }
        int j = 0;
        ListNode curr = pre.next;
        while (j < right - left){
            ListNode reserve = curr.next;
            curr.next = reserve.next;
            reserve.next = pre.next;
            pre.next = reserve;
            j++;
        }
        return dummy.next;
    }

    /**
     *  反转指定节点之间的节点
     */
    public ListNode[] reverseBetween(ListNode head, ListNode tail){
        ListNode prev = tail.next;
        ListNode p = head;
        while (prev != tail){
            ListNode next = p.next;
            p.next = prev;
            prev = p;
            p = next;
        }
        return new ListNode[]{tail, head};
    }

    /**
     *  使用 head 和 tail 指针指向需要反转的位置，如果长度小于 k 则不反转
     */
    public ListNode reverseKGroup(ListNode head, int k){

        ListNode dummy = new ListNode(-1, head);
        ListNode prev = dummy;

        while (head != null){

            ListNode tail = prev;
            for (int i = 0; i < k; i++){
                tail = tail.next;
                if (tail == null){
                    return dummy.next;
                }
            }

            ListNode next = tail.next;
            ListNode[] reverse =reverseBetween(head, tail);
            head = reverse[0];
            tail = reverse[1];

            // 拼接
            prev.next = head;
            tail.next = next;

            prev = tail;
            head = tail.next;
        }
        return dummy.next;
    }
}
